packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/data_query/data_query.rs

```diff
@@ -89,43 +101,67 @@ impl From<LocalPropIdx> for PropSpecifier {
-    /// Returns the matching props
-    ChildPropProfile {
-        /// The data query will match child components that have at least one of these profiles
-        match_profiles: Vec<PropProfile>,
-    },
-
-    /// Query for all child GraphNodes,
-    /// filtering element nodes for components which contain PropProfile props given specified values
+    /// Query for all child GraphNodes, filtering element nodes based on the supplied
+    /// `filter`.
     ///
-    /// filter: a vector of `PropProfile` and `PropValue` pairings.
-    /// If a component has a prop with each specified `PropProfile`,
-    /// then it will be filtered out if any of the values don't match the specified `PropValue`.
-    /// If a component does not have a prop with one or more of the specified `PropProfiles`,
-    /// then it will be included if `include_if_missing_profile` is `true`, otherwise it will be be filtered out.
-    /// All non-component children are unaffected by the filter.
+    /// - `container`: the component whose children will be searched. For example, `PropComponent::Me`
+    /// to search your own children, or `PropComponent::Parent` to search your parent's children.
+    /// - `filter`: A composition of `ContentFilter`s. These can be combined with `Op::And`, `Op::Or`,
+    /// and `OpNot`.
     ///
-    /// For example, if the filter has one element `(PropProfile:Hidden`, `PropValue::Boolean(false))`, then
-    /// - if `include_if_missing_profile` is `true`,
-    ///   all component children that have a prop with the profile `Hidden` whose value is not `false`
-    ///   will be filtered out.
-    /// - if `include_if_missing_profile` is `false`,
-    ///   all component children that do not have a prop with the profile `Hidden` with value `false`
-    ///   will be filtered out.
-    FilteredChildren {
-        filters: Vec<DataQueryFilter>,
-        include_if_missing_profile: bool,
+    /// ## Example
+    /// ```rust
+    /// # use std::rc::Rc;
+    /// # use doenetml_core::props::{DataQuery, PropSource, ContentFilter, Op, PropProfile};
+    /// DataQuery::ComponentRefs {
+    ///    container: PropSource::Me,
+    ///    filter: Rc::new(Op::And(
+    ///      ContentFilter::IsType("section"),
+    ///      ContentFilter::HasPropMatchingProfile(PropProfile::Hidden),
+    ///    ))
+    /// }
+    /// # ;
+    /// ```
+    /// This query will return all children of the querying component that are `<section>` components
+    /// and have a prop matching the `PropProfile::Hidden` profile.
+    ComponentRefs {
+        /// Children of this component will be searched
+        container: PropSource,
+        /// How to filter the children. This should be a [`ContentFilter`] or a
+        /// composition of [`ContentFilter`]s.
+        ///
+        /// See [`DataQuery::ComponentRefs`] for an example.
+        filter: Rc<dyn for<'a> BindableAsGraphNodeFilter<'a>>,
```

Comment:
I guess the fact that this `filter` is so different from the `prop_specifier` is an argument for having a different mechanism for the cases where we aren't grabbing a prop. But, from a high level, it seems like the cases with or without selecting a prop should be similar.

I still haven't done the work to figure out how a `BindableAsGraphNodeFilter` works. Have you decided whether or not we should remove one level of dynamic trait objects?

Reply:
I think it's fine to keep. It's all working, and does make the usage nice. The only non-nice thing is that `BindableAsGrahNodeFilter` doesn't tell you that you should be using `Op::...` to build your query.